packages/@react-stately/select/src/useSelectState.ts

```diff
@@ -44,34 +78,85 @@ export interface SelectState<T> extends SingleSelectListState<T>, OverlayTrigger
  * of items from props, handles the open state for the popup menu, and manages
  * multiple selection state.
  */
-export function useSelectState<T extends object>(props: SelectStateOptions<T>): SelectState<T>  {
+export function useSelectState<T extends object, M extends SelectionMode = 'single'>(props: SelectStateOptions<T, M>): SelectState<T, M>  {
+  let {selectionMode = 'single' as M} = props;
   let triggerState = useOverlayTriggerState(props);
   let [focusStrategy, setFocusStrategy] = useState<FocusStrategy | null>(null);
-  let listState = useSingleSelectListState({
+  let defaultValue = useMemo(() => {
+    return props.defaultValue ?? (selectionMode === 'single' ? props.defaultSelectedKey ?? null : []) as ValueType<M>;
+  }, [props.defaultValue, props.defaultSelectedKey, selectionMode]);
+  let value = useMemo(() => {
+    return props.value ?? (selectionMode === 'single' ? props.selectedKey : undefined) as ValueType<M>;
+  }, [props.value, props.selectedKey, selectionMode]);
+  let [controlledValue, setControlledValue] = useControlledState<ValueType<M>>(value as any, defaultValue as any, props.onChange);
+  let setValue = (value: Key | Key[] | null) => {
+    if (selectionMode === 'single') {
+      let key = Array.isArray(value) ? value[0] ?? null : value;
+      setControlledValue(key as ValueType<M>);
+      if (key !== controlledValue) {
+        props.onSelectionChange?.(key);
+      }
+    } else {
+      let keys: Key[] = [];
+      if (Array.isArray(value)) {
+        keys = value;
+      } else if (value != null) {
+        keys = [value];
+      }
+
+      setControlledValue(keys as ValueType<M>);
+    }
+  };
+
+  let listState = useListState({
     ...props,
-    onSelectionChange: (key) => {
-      if (props.onSelectionChange != null) {
-        props.onSelectionChange(key);
+    selectionMode,
+    disallowEmptySelection: selectionMode === 'single',
+    allowDuplicateSelectionEvents: true,
+    selectedKeys: useMemo(() => convertValue(controlledValue), [controlledValue]),
+    onSelectionChange: (keys: Selection) => {
+      // impossible, but TS doesn't know that
+      if (keys === 'all') {
+        return;
+      }
+
+      if (selectionMode === 'single') {
+        let key = keys.values().next().value ?? null;
+        setValue(key);
+        triggerState.close();
```

Comment:
combine work to keep open with https://github.com/adobe/react-spectrum/pull/8733 ?

Reply:
yeah we will but we can merge them separately